# Single-cell multi-omics profiling links dynamic DNA methylation to cell fate decisions during early mammalian organogenesis

This repository contains the scripts to reproduce the Tet-TKO scNMT-seq analysis of [the manuscript](XXX). It covers Figure 4 as well as Supplementary Figures 7-8.
The source code for the Tet-TKO scNMT-seq, is available in [this repository](https://github.com/rargelaguet/scnmt_gastrulation_TetChimera)


<p align="center"> 
<img src="images/fig.png" width="850" height="550"/>
</p>


Experimental design
--------
XXXX


Key results
--------

- XXX

- YYY

- ZZZ

Data
-------
The data is currently not publicly available, we are working on this...

<!-- The raw data is accessible at GEO ([GSEXXXX](XXXX)). 
The parsed data can be downloaded [here](XXXX) -->


Shiny app
--------
We provide a shiny app that allows the user to explore the Tet-TKO scRNA-seq results [here](https://www.bioinformatics.babraham.ac.uk/shiny/tet_ko_embryo_scrna/). Note that we currently do not have a shiny app to explore the scNMT-seq TET-TKO results.

Contact
-------

For questions on the computational analysis: Ricard Argelaguet (ricard.argelaguet@gmail.com). For questions on the experimental work: Tim Lohoff (tlohoff431@gmail.com) or Stephen Clark (Stephen.Clark@babraham.ac.uk)

